From: Sebastian Ott <sebott@linux.vnet.ibm.com>
Subject: s390/pci: implement pcibios_add_device
Patch-mainline: v3.10-rc1
Git-commit: af0a8a8453f7c7b3497c9fecc053897690e00695
References: bnc#848335,FATE#83037,LTC#94737

Use pcibios_add_device to do arch specific device initialization.
This function will be called during pci_bus_add_device.

Reviewed-by: Gerald Schaefer <gerald.schaefer@de.ibm.com>
Signed-off-by: Sebastian Ott <sebott@linux.vnet.ibm.com>
Signed-off-by: Martin Schwidefsky <schwidefsky@de.ibm.com>
Acked-by: John Jolly <jjolly@suse.de>
---
 arch/s390/pci/pci.c |   14 +++++++++++---
 1 file changed, 11 insertions(+), 3 deletions(-)

--- a/arch/s390/pci/pci.c
+++ b/arch/s390/pci/pci.c
@@ -867,6 +867,17 @@ static void zpci_free_iomap(struct zpci_
 	spin_unlock(&zpci_iomap_lock);
 }
 
+int pcibios_add_device(struct pci_dev *pdev)
+{
+	struct zpci_dev *zdev = get_zdev(pdev);
+
+	zpci_debug_init_device(zdev);
+	zpci_fmb_enable_device(zdev);
+	zpci_map_resources(zdev);
+
+	return 0;
+}
+
 static int zpci_create_device_bus(struct zpci_dev *zdev)
 {
 	struct resource *res;
@@ -1016,9 +1027,6 @@ int zpci_scan_device(struct zpci_dev *zd
 		goto out;
 	}
 
-	zpci_debug_init_device(zdev);
-	zpci_fmb_enable_device(zdev);
-	zpci_map_resources(zdev);
 	pci_bus_add_devices(zdev->bus);
 
 	/* now that pdev was added to the bus mark it as used */
